casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -86,55 +101,46 @@ object EstimatorHelper {
-              (produces :+ produce, consumes, commEvents)
-            case Event(Consume(consume: ConsumeEvent)) =>
-              (produces, consumes :+ consume, commEvents)
-            case Event(Comm(commEvent: CommEvent)) =>
-              (produces, consumes, commEvents :+ commEvent)
-            case _ =>
-              throw new RuntimeException("Unexpected ancestor event")
-          }
-      }
-      producesInCommEvents = commEvents.flatMap {
-        case CommEvent(Some(_: ConsumeEvent), produces) =>
-          produces
-        case _ =>
-          throw new RuntimeException("Unexpected comm event")
-      }
-      consumeInCommEvents = commEvents.map {
-        case CommEvent(Some(consume: ConsumeEvent), _) =>
-          consume
-        case _ =>
-          throw new RuntimeException("Unexpected comm event")
-      }
-      ancestorChannels = ancestorEvents.flatMap {
-        case Event(Produce(produce: ProduceEvent)) =>
-          if (producesInCommEvents.contains(produce)) {
-            Set.empty[ByteString] // Volatile produce
-          } else {
-            Set(produce.channelsHash)
-          }
-        case Event(Consume(consume: ConsumeEvent)) =>
-          if (consumeInCommEvents.contains(consume)) {
-            Set.empty[ByteString] // Volatile consume
-          } else {
-            consume.channelsHashes.toSet
-          }
-        case Event(Comm(CommEvent(Some(consume: ConsumeEvent), produces))) =>
-          if (consumeEvents.contains(consume) && produces.forall(
-                produce => produceEvents.contains(produce)
-              )) {
-            Set.empty[ByteString] // Volatile consume & produces
-          } else {
-            consume.channelsHashes.toSet ++ produces.map(_.channelsHash).toSet
-          }
-        case _ => throw new RuntimeException("incorrect ancestor events")
-      }.toSet
+
+      produceEvents = ancestorEvents.map(extractProduce).flatten
+      consumeEvents = ancestorEvents.map(extractConsume).flatten
+      commEvents    = ancestorEvents.map(extractComm).flatten
+
+      producesInCommEvents = commEvents.flatMap(_.produces)
+      consumeInCommEvents = commEvents.flatMap(_.consume)
+
+      ancestorProduceChannels = ancestorEvents
+        .map(extractProduce)
+        .flatten
+        .filterNot(producesInCommEvents.contains(_))
+        .map(p => p.channelsHash)
+        .toSet
+
+      ancestorConsumeChannels = ancestorEvents
+        .map(extractConsume)
+        .flatten
+        .filterNot(consumeInCommEvents.contains(_))
+        .flatMap(c => c.channelsHashes)
+        .toSet
+
+      ancestorCommChannels = ancestorEvents
+        .map(extractComm)
+        .flatten
+        .filterNot {
+          case CommEvent(Some(consume: ConsumeEvent), produces) =>
+            consumeEvents.contains(consume) && produces.forall(
+              produce => produceEvents.contains(produce)
+            )
+          case _ => false
```

Comment:
please extract `isVolatile(commEvent: CommEvent) and use it here.

The `case _ =>` should be throwing an exception.